src/v/storage/parser.cc

```diff
@@ -93,15 +93,19 @@ static ss::future<result<iobuf>> verify_read_iobuf(
 }
 
 ss::future<result<stop_parser>> continuous_batch_parser::consume_header() {
-    return read_iobuf_exactly(_input, model::packed_record_batch_header_size)
+    auto f = ss::make_ready_future<result<model::record_batch_header>>(_header);
+    if (_header.header_crc == 0 && _header.size_bytes == 0) {
+      f = read_iobuf_exactly(_input, model::packed_record_batch_header_size)
```

Comment:
this logic here is hard to read. should we refactor this with coros and then add this logic. 

at the very least we should add a comment.

Reply:
with coros it looks much better.